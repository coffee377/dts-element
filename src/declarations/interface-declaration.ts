import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {transform} from '../transform';
import {IGeneralType} from '../types/general-type';
import {create_object_type, IObjectType} from '../types/object-type';
import {add_declare_modifier_if_need, create_type_arguments, create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface IInterfaceDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  type?: IObjectType;
  extends?: IGeneralType[];
}

export interface IInterfaceDeclaration
  extends IElement<ElementKind.InterfaceDeclaration>, IInterfaceDeclarationOptions {}

export const create_interface_declaration = (options: IInterfaceDeclarationOptions): IInterfaceDeclaration => ({
  ...create_element(ElementKind.InterfaceDeclaration),
  ...options,
});

// tslint:disable:ter-indent max-line-length

export const transform_interface_declaration = (element: IInterfaceDeclaration, path: IElement<any>[]) =>
  ts.createInterfaceDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
                            (element.export === true)
                              ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                              : undefined,
                            path,
                          ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ element.extends && [ts.createHeritageClause(
                            /* token */ ts.SyntaxKind.ExtendsKeyword,
                            /* types */ element.extends.map(interface_type => ts.createExpressionWithTypeArguments(
                                          /* typeArguments */ create_type_arguments(interface_type.generics, path),
                                          /* expression    */ ts.createIdentifier(interface_type.name),
                                        )),
                          )],
    /* members         */ (transform(element.type || create_object_type(), path) as ts.TypeLiteralNode).members,
  );
