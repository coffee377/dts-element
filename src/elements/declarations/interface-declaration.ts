import {emit_export} from '../../helpers/emit-export';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_members} from '../../helpers/emit-members';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ObjectMember} from '../members/object-member';
import {GenericType} from '../types/generic-type';

export interface IInterfaceDeclarationRequiredParameters {
  name: string;
}

export interface IInterfaceDeclarationOptionalParameters {
  generics: GenericType[];
  members: ObjectMember[];
  export: boolean;
}

/**
 * ```ts
 * interface some_interface_name {
 *   // some object members
 * }
 * ```
 */
export class InterfaceDeclaration
    extends Declaration<IInterfaceDeclarationRequiredParameters, IInterfaceDeclarationOptionalParameters> {

  private _instance_of_interface_declaration: true;

  public get default_parameters(): IDeclarationOptionalParameters & IInterfaceDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      members: [],
      export: false,
    });
  }

  public _emit_raw(stack: Stack): string {
    const {name, members, generics} = this.parameters;
    const generic = emit_generics(generics, stack, true);
    const an_export = emit_export(this.parameters.export, stack);
    return `${an_export}interface ${name}${generic} ${emit_members(members, stack)}`;
  }

}
