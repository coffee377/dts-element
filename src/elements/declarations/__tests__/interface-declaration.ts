jest.unmock('../interface-declaration');

import {InterfaceDeclaration} from '../interface-declaration';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new InterfaceDeclaration({name: 'A'}).emit()).toMatchSnapshot();
  });
});
