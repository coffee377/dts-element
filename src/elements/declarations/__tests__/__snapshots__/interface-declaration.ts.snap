// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"[emit-jsdoc][emit-export]interface A[emit-generics] [emit-members]"`;
